### ISSUE TITLE
Is this useful?

### ISSUE THREAD

**Issue by abelsiqueira** [abelsiqueira/NLPClass.jl#1]
@dpo, @mlubin. Before I start wasting too much time on this idea, please tell me what you think of this packages. Thanks.

This package is supposed to help create a classification of problems on-the-fly. It works like this:
````julia
@NLPClassify(class, Rosenbrock,
  begin
    nvar = 2
    obj = sos
    con = unc
    open = rosen
    model_type = JuMPModel
  end)
function rosen()
  nlp = Model()
  x0 = [-1.2; 1.0]
  @variable(nlp, x[i=1:2], start=x0[i])
  @NLobjective(nlp, Min, 1.0 + (x[1] - 1.0)^2 + 100.0 * (x[2] - x[1]^2)^2)

  return nlp
end
````
or for CUTEst
````julia
@NLPClassify(class, ROSENBR,
  begin
    nvar = 2
    obj = sos
    con = unc
    open = CUTEstModel("ROSENBR")
    model_type = AbstractNLPModel
  end
)
# ROSENBR is defined at CUTEst
````
## Advantages

- It's very general, so it can be used with any problem type. Although, here I'm focusing on [CUTEst](https://github.com/JuliaSmoothOptimizers/CUTEst.jl) and [OptimizationProblems.jl](https://github.com/JuliaSmoothOptimizers/OptimizationProblems.jl).
- New categories can be added without breaking existing ones. E.g., "convex", "bounded", "best_f".
- Unified access:
````julia
for problem in keys(class.entries)
  println("Opening problem $problem")
  t = getType(class, problem)
  nlp = eval(open(class, problem))
  if t == "JuMPModel"
    nlp = MathProgNLPModel(eval(open(class, problem)))
  elseif t != "AbstractNLPModel"
    error("Unexpected type $t")
  end
  println("  x0 = $(nlp.meta.x0)")
  println("  f(x0) = $(obj(nlp, nlp.meta.x0))")
  finalize(nlp)
end
````

## Disavantages

- It's probably slow. Haven't tested how slow yet.
- Creates overhead
- It may be complicated.

## Alternatives

- Static JSON file, like what is currently done in CUTEst (undocumented).

**Comment by mlubin** [comment]
CC @pkofod

**Comment by abelsiqueira** [comment]
CC @timholy

**Comment by timholy** [comment]
Yes, it's very useful to have problem classifications. Of course, as you say the API is likely to be tricky to support all needs. See also https://github.com/JuliaOpt/Optim.jl/pull/303/commits/dcfe788c2fc775c967b64acc82fe85e5e983c89f for recent (unmerged) related work.

**Comment by pkofod** [comment]
Could potentially be quite useful for refining insights from benchmarks.
